studio/src/lib/schema-helpers.ts

```diff
@@ -656,9 +685,84 @@ export const getDeprecatedTypes = (
   return deprecatedTypes;
 };
 
+type ExtractedDirectives = {
+  authenticated: boolean;
+  requiresScopes: string[][] | null;
+  deprecationReason: string;
+}
+
+const extractDirectives = (
+    astNode: ASTNode | undefined | null
+) => {
+  const result: ExtractedDirectives = {
+    authenticated: false,
+    requiresScopes: null,
+    deprecationReason: '',
+  };
+  
+  if (!astNode) {
+    return result;
+  }
+
+  visit(astNode, {
+    Directive(node) {
+      if (node.name.value === "deprecated") {
+        const arg = node.arguments?.[0]?.value;
+        if (arg?.kind === Kind.STRING) {
+          result.deprecationReason = arg.value;
+        }
+      } else if (node.name.value === "authenticated") {
+        result.authenticated = true;
+      } else if (node.name.value === "requiresScopes" && Array.isArray(node.arguments)) {
+        result.requiresScopes = node.arguments
+            .filter((arg: ConstArgumentNode) => arg.name.value === "scopes" && arg.value.kind === Kind.LIST)
+            .flatMap((arg: ConstArgumentNode) => (arg.value as ListValueNode).values)
+            .map((value) => (value as ListValueNode).values.map((sv) => (sv as StringValueNode).value));
+      }
+    },
+  });
+
+  return result;
+}
+
+export const getAuthenticatedTypes = (
```

Comment:
You could have a `FieldDefinitionVisitor`, and if the field starts with `__`, you can return false (it will immediately leave and not go deeper on this level).

You can then handle everything though a directive visitor and its ancestors? Then you have a single visitor that visits exactly once, and could handle all relevant directives. But for now, perhaps just handle authenticated and requiresScopes, and we can refactor deprecated later.

What do you think?

Reply:
I think that sounds good, also, with `FieldDefinitionVisitor` you mean an `ObjectType` visitor right? If the name starts with `__` bail

Reply:
@wilsonrivera outcome?

Reply:
Sorry for the delay, I've implemented it two steps, one to visit the field and one to visit the directives.